transactions/src/main/java/org/jboss/as/txn/service/LocalTransactionContextService.java

```diff
@@ -50,6 +50,7 @@ public final class LocalTransactionContextService implements Service<LocalTransa
     private final InjectedValue<XAResourceRecoveryRegistry> xaResourceRecoveryRegistryInjector = new InjectedValue<>();
     private final InjectedValue<ServerEnvironment> serverEnvironmentInjector = new InjectedValue<>();
     private final int staleTransactionTime;
+    private JBossLocalTransactionProvider provider;
```

Comment:
is `provider` thread-safe? As a data point in line 47 `context` is declared `volatile`, and both fields seems to have similar purpose. Or can we not add `provider`, and get the provider from `this.context` in `stop` method?

Reply:
Thanks for the review @chengfang .  Good point!
What about setting start() and stop() method as synchronized instead?

Reply:
Ok, but there is no getter in LocalTransactionContext for the provider, and that is private (see https://github.com/wildfly/wildfly-transaction-client/blob/master/src/main/java/org/wildfly/transaction/client/LocalTransactionContext.java#L71). I don't know exactly how to "get the provider from this.context in stop method"?

Reply:
sorry, I missed that. I really wished it could be public. Then how about making `provider` volatile too?

Reply:
I am not against using `volatile` on the `provider`. If in the stop() method there was the `this.provider = null` assignment I would agree. But here in the stop method the provider reference does not change.
IMO if we need to have the provider more thread-safe we would better execute the start and the stop method not in parallel (i.e. with synchronized methods) .

Reply:
Okay, that's good point. I think what we have should be good.

Reply:
Note that MSC won't invoke start and stop concurrently.